File: appkit/stack_view.m
#import <Appkit/NSStackView.h>
#import "stack_view.h"

Array StackView_Views(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	NSArray* array = [stackView views];
	int count = [array count];
	void** data = malloc(count * sizeof(void*));
	for (int i = 0; i < count; i++) {
		 data[i] = [array objectAtIndex:i];
	}
	Array result;
	result.data = data;
	result.len = count;
	return result;
}

Array StackView_DetachedViews(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	NSArray* array = [stackView detachedViews];
	int count = [array count];
	void** data = malloc(count * sizeof(void*));
	for (int i = 0; i < count; i++) {
		 data[i] = [array objectAtIndex:i];
	}
	Array result;
	result.data = data;
	result.len = count;
	return result;
}

long StackView_Orientation(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView orientation];
}

void StackView_SetOrientation(void* ptr, long orientation) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setOrientation:orientation];
}

long StackView_Alignment(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView alignment];
}

void StackView_SetAlignment(void* ptr, long alignment) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setAlignment:alignment];
}

double StackView_Spacing(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView spacing];
}

void StackView_SetSpacing(void* ptr, double spacing) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setSpacing:spacing];
}

NSEdgeInsets StackView_EdgeInsets(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView edgeInsets];
}

void StackView_SetEdgeInsets(void* ptr, NSEdgeInsets edgeInsets) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setEdgeInsets:edgeInsets];
}

long StackView_Distribution(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView distribution];
}

void StackView_SetDistribution(void* ptr, long distribution) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setDistribution:distribution];
}

Array StackView_ArrangedSubviews(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	NSArray* array = [stackView arrangedSubviews];
	int count = [array count];
	void** data = malloc(count * sizeof(void*));
	for (int i = 0; i < count; i++) {
		 data[i] = [array objectAtIndex:i];
	}
	Array result;
	result.data = data;
	result.len = count;
	return result;
}

bool StackView_DetachesHiddenViews(void* ptr) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView detachesHiddenViews];
}

void StackView_SetDetachesHiddenViews(void* ptr, bool detachesHiddenViews) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setDetachesHiddenViews:detachesHiddenViews];
}

void* StackView_StackViewWithViews(Array views) {
    NSMutableArray* objcViews = [[NSMutableArray alloc] init];
    void** viewsData = (void**)views.data;
    for (int i = 0; i < views.len; i++) {
    	[objcViews addObject:(NSView*)viewsData[i]];
    }
	return [NSStackView stackViewWithViews:objcViews];
}

Array StackView_ViewsInGravity(void* ptr, long gravity) {
	NSStackView* stackView = (NSStackView*)ptr;
	NSArray* array = [stackView viewsInGravity:gravity];
	int count = [array count];
	void** data = malloc(count * sizeof(void*));
	for (int i = 0; i < count; i++) {
		 data[i] = [array objectAtIndex:i];
	}
	Array result;
	result.data = data;
	result.len = count;
	return result;
}

void StackView_AddView(void* ptr, void* view, long gravity) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView addView:(NSView*)view inGravity:gravity];
}

void StackView_InsertView(void* ptr, void* view, unsigned long index, long gravity) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView insertView:(NSView*)view atIndex:index inGravity:gravity];
}

void StackView_SetViews(void* ptr, Array views, long gravity) {
	NSStackView* stackView = (NSStackView*)ptr;
    NSMutableArray* objcViews = [[NSMutableArray alloc] init];
    void** viewsData = (void**)views.data;
    for (int i = 0; i < views.len; i++) {
    	[objcViews addObject:(NSView*)viewsData[i]];
    }
	[stackView setViews:objcViews inGravity:gravity];
}

void StackView_RemoveView(void* ptr, void* view) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView removeView:(NSView*)view];
}

void StackView_AddArrangedSubview(void* ptr, void* view) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView addArrangedSubview:(NSView*)view];
}

void StackView_InsertArrangedSubview(void* ptr, void* view, unsigned long index) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView insertArrangedSubview:(NSView*)view atIndex:index];
}

void StackView_RemoveArrangedSubview(void* ptr, void* view) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView removeArrangedSubview:(NSView*)view];
}

float StackView_ClippingResistancePriorityForOrientation(void* ptr, long orientation) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView clippingResistancePriorityForOrientation:orientation];
}

void StackView_SetClippingResistancePriority(void* ptr, float clippingResistancePriority, long orientation) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setClippingResistancePriority:clippingResistancePriority forOrientation:orientation];
}

float StackView_HuggingPriorityForOrientation(void* ptr, long orientation) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView huggingPriorityForOrientation:orientation];
}

void StackView_SetHuggingPriority(void* ptr, float huggingPriority, long orientation) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setHuggingPriority:huggingPriority forOrientation:orientation];
}

double StackView_CustomSpacingAfterView(void* ptr, void* view) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView customSpacingAfterView:(NSView*)view];
}

void StackView_SetCustomSpacing(void* ptr, double spacing, void* view) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setCustomSpacing:spacing afterView:(NSView*)view];
}

float StackView_VisibilityPriorityForView(void* ptr, void* view) {
	NSStackView* stackView = (NSStackView*)ptr;
	return [stackView visibilityPriorityForView:(NSView*)view];
}

void StackView_SetVisibilityPriority(void* ptr, float priority, void* view) {
	NSStackView* stackView = (NSStackView*)ptr;
	[stackView setVisibilityPriority:priority forView:(NSView*)view];
}

File: scripts/stack_view.py
#!env python3

from generate import Component, Property, init_method, Param, Method, Return

if __name__ == "__main__":
    w = Component(
        Type="appkit.StackView",
        super_type='appkit.View',
        description="a view that arranges an array of views horizontally or vertically and updates their placement and sizing when the window size changes",
        properties=[
            Property(name='views', Type='appkit.View', readonly=True, array=True,
                     description='the array of views owned by the stack view'),
            Property(name='detachedViews', Type='appkit.View', readonly=True, array=True,
                     description='the detached views from all the stack view’s gravity areas'),
            Property(name='orientation', Type='int', go_alias_type='UserInterfaceLayoutOrientation',
                     description='the horizontal or vertical layout direction of the stack view'),
            Property(name='alignment', Type='int', go_alias_type='LayoutAttribute',
                     description='the view alignment within the stack view'),
            Property(name='spacing', Type='float64',
                     description='the minimum spacing, in points, between adjacent views in the stack view'),
            Property(name='edgeInsets', Type='foundation.EdgeInsets',
                     description='the geometric padding, in points, inside the stack view, surrounding its views'),
            Property(name='distribution', Type='int', go_alias_type='StackViewDistribution',
                     description='distribution'),
            Property(name='arrangedSubviews', Type='appkit.View', readonly=True, array=True,
                     description='the array of views arranged by the stack view'),
            Property(name='detachesHiddenViews', Type='bool', getter_prefix_is=False,
                     description='whether the stack view removes hidden views from its view hierarchy'),

        ],
        methods=[
            Method(
                name='stackViewWithViews',
                register_delegate=True,
                static=True,
                params=[Param(name='views', Type='appkit.View', array=True)],
                return_value=Return(Type='appkit.StackView'),
                description='return a stack view with views',
            ),
            Method(
                name='viewsInGravity',
                params=[Param(name='gravity', Type='int', go_alias='StackViewGravity')],
                return_value=Return(Type='appkit.View', array=True),
                description='returns the array of views in the specified gravity area in the stack view',
            ),
            Method(
                name='addView',
                params=[
                    Param(name='view', Type='appkit.View'),
                    Param(name='gravity', Type='int', go_alias='StackViewGravity', objc_param_name='inGravity'),
                ],
                description='adds a view to the end of the stack view gravity area',
            ),
            Method(
                name='insertView',
                params=[
                    Param(name='view', Type='appkit.View'),
                    Param(name='index', Type='uint', objc_param_name='atIndex'),
                    Param(name='gravity', Type='int', go_alias='StackViewGravity', objc_param_name='inGravity'),
                ],
                description='adds a view to a stack view gravity area at a specified index position',
            ),
            Method(
                name='setViews',
                params=[
                    Param(name='views', Type='appkit.View', array=True),
                    Param(name='gravity', Type='int', go_alias='StackViewGravity', objc_param_name='inGravity'),
                ],
                description='specifies an array of views for a specified gravity area in the stack view, replacing any previous views in that area',
            ),
            Method(
                name='removeView',
                params=[
                    Param(name='view', Type='appkit.View'),
                ],
                description='removes a specified view from the stack view',
            ),
            Method(
                name='addArrangedSubview',
                params=[
                    Param(name='view', Type='appkit.View'),
                ],
                description='adds the specified view to the end of the arranged subviews list',
            ),
            Method(
                name='insertArrangedSubview',
                params=[
                    Param(name='view', Type='appkit.View'),
                    Param(name='index', Type='uint', objc_param_name='atIndex'),
                ],
                description='adds the provided view to the array of arranged subviews at the specified index',
            ),
            Method(
                name='removeArrangedSubview',
                params=[
                    Param(name='view', Type='appkit.View'),
                ],
                description='removes the provided view from the stack’s array of arranged subviews',
            ),
            Method(
                name='clippingResistancePriorityForOrientation',
                params=[
                    Param(name='orientation', Type='int', go_alias='LayoutConstraintOrientation'),
                ],
                return_value=Return(Type='float32', go_alias='LayoutPriority'),
                description='returns the Auto Layout priority for resisting clipping of views in the stack view when Auto Layout attempts to reduce the stack view’s size',
            ),
            Method(
                name='setClippingResistancePriority',
                params=[
                    Param(name='clippingResistancePriority', Type='float32', go_alias='LayoutPriority'),
                    Param(name='orientation', Type='int', go_alias='LayoutConstraintOrientation',
                          objc_param_name='forOrientation'),
                ],
                description='sets the Auto Layout priority for resisting clipping of views in the stack view when Auto Layout attempts to reduce the stack view’s size',
            ),
            Method(
                name='huggingPriorityForOrientation',
                params=[
                    Param(name='orientation', Type='int', go_alias='LayoutConstraintOrientation'),
                ],
                return_value=Return(Type='float32', go_alias='LayoutPriority'),
                description='returns the Auto Layout priority for the stack view to minimize its size to fit its contained views as closely as possible, for a specified user interface axis',
            ),
            Method(
                name='setHuggingPriority',
                params=[
                    Param(name='huggingPriority', Type='float32', go_alias='LayoutPriority'),
                    Param(name='orientation', Type='int', go_alias='LayoutConstraintOrientation',
                          objc_param_name='forOrientation'),
                ],
                description='sets the Auto Layout priority for the stack view to minimize its size, for a specified user interface axis',
            ),
            Method(
                name='customSpacingAfterView',
                params=[
                    Param(name='view', Type='appkit.View'),
                ],
                return_value=Return(Type='float64'),
                description='returns the custom spacing, in points, between a specified view in the stack view and the view that follows it',
            ),
            Method(
                name='setCustomSpacing',
                params=[
                    Param(name='spacing', Type='float64'),
                    Param(name='view', Type='appkit.View', objc_param_name='afterView'),
                ],
                description='specifies the custom spacing, in points, between a specified view and the view that follows it in the stack view',
            ),
            Method(
                name='visibilityPriorityForView',
                params=[
                    Param(name='view', Type='appkit.View'),
                ],
                return_value=Return(Type='float32', go_alias='StackViewVisibilityPriority'),
                description='returns the visibility priority for a specified view in the stack view',
            ),
            Method(
                name='setVisibilityPriority',
                params=[
                    Param(name='priority', Type='float32', go_alias='StackViewVisibilityPriority'),
                    Param(name='view', Type='appkit.View', objc_param_name='forView'),
                ],
                description='sets the Auto Layout priority for a view to remain attached to the stack view when Auto Layout reduces the stack view’s size',
            ),

        ]
    )
    w.generate_code()
